olive/passes/onnx/float32_conversion.py

```diff
@@ -0,0 +1,126 @@
+# -------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# --------------------------------------------------------------------------
+import logging
+import re
+from collections import defaultdict
+from pathlib import Path
+from typing import Any, Dict
+
+import onnx
+
+from olive.hardware.accelerator import AcceleratorSpec
+from olive.model import ONNXModelHandler
+from olive.model.utils import resolve_onnx_path
+from olive.passes import Pass
+from olive.passes.onnx.common import get_external_data_config, model_proto_to_olive_model
+from olive.passes.pass_config import PassConfigParam
+
+logger = logging.getLogger(__name__)
+
+
+class OnnxIOFloat16ToFloat32(Pass):
+    """Converts float16 model inputs/outputs to float32."""
+
+    @classmethod
+    def _default_config(cls, accelerator_spec: AcceleratorSpec) -> Dict[str, PassConfigParam]:
+        config = {
+            "name_pattern": PassConfigParam(
+                type_=str,
+                default_value="logits",
+                description=(
+                    "Only convert inputs/outputs whose name matches this pattern. By defaultlooking for logits names"
+                ),
+            )
+        }
+        config.update(get_external_data_config())
+        return config
+
+    def create_io_mapping(self, graph, i_map, o_map):
+        for n in graph.node:
+            for i in n.input:
+                i_map[i].append(n)
+        for n in graph.node:
+            for o in n.output:
+                assert o not in o_map
+                o_map[o] = [n]
+
+    def wrap_inputs(self, graph, i_map, names):
+        # 1. find fp16 inputs
+        # 2. rewrite all consumers
+        # 3. insert cast
+        # 4. rewrite graph inputs
+        inputs = [n for n in graph.input if n.type.tensor_type.elem_type == onnx.TensorProto.FLOAT16]
+        for i in inputs:
+            if names:
+                match = names.search(i.name)
+                if not match:
+                    continue
+            logger.debug("input %s from fp32", i.name)
+            for n in i_map[i.name]:
+                for j, o in enumerate(n.input):
+                    if o == i.name:
+                        n.input[j] = i.name + "_fp16"
+            cast = onnx.helper.make_node(
+                "Cast",
+                inputs=[i.name],
+                outputs=[i.name + "_fp16"],
+                to=onnx.TensorProto.FLOAT,
```

Comment:
to fp32 - javascript doesn't know what fp16 is

Reply:
Since this is a cast in the input from `{i.name}` to `{i.name}_fp16`, the cast to `FLOAT` did not make sense to me. We are essentially casting from `float` to `float` with this node since `{i.name}` is changed to float in line 72.

Reply:
My understanding is we want: `fp32 input` -> `cast to fp16` -> `Graph` -> `cast to fp32` ->`fp32 input`. In which case this cast `to` must be `FLOAT16`?

Reply:
yes

Reply:
In this case. I think line 54 should select FLOAT rather FLOAT16

Reply:
line 54 is fine.